From d766c3d37a27b785ffa5f9bda09a5ee5c9633da2 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Mon, 5 May 2014 19:52:43 -0700
Subject: [PATCH 094/456] drm: xilinx: plane: Add the format property in the
 planes node

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

Add the format property in the planes node. This allows to configure
the pixel format properly as the format can differ depending on configuration.

Signed-off-by: Hyun Kwon <hyunk@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit 5bd367d826c57bc12b9a645a63d3fce3b9ab3486)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/xilinx/xilinx_drm_plane.c |   22 +++++++++++++++++++---
 1 files changed, 19 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_plane.c b/drivers/gpu/drm/xilinx/xilinx_drm_plane.c
index ff05b4c..bbaf9d5 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_plane.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_plane.c
@@ -813,6 +813,7 @@ static int
 xilinx_drm_plane_init_manager(struct xilinx_drm_plane_manager *manager)
 {
 	unsigned int format;
+	uint32_t drm_format;
 	int ret = 0;
 
 	if (manager->osd) {
@@ -820,12 +821,12 @@ xilinx_drm_plane_init_manager(struct xilinx_drm_plane_manager *manager)
 		manager->max_width = xilinx_osd_get_max_width(manager->osd);
 
 		format = xilinx_osd_get_format(manager->osd);
-		ret = xilinx_drm_format_by_code(format, &manager->format);
+		ret = xilinx_drm_format_by_code(format, &drm_format);
+		if (drm_format != manager->format)
+			ret = -EINVAL;
 	} else {
 		/* without osd, only one plane is supported */
 		manager->num_planes = 1;
-		/* YUV422 based on the current pipeline design without osd */
-		manager->format = DRM_FORMAT_YUYV;
 		manager->max_width = 4096;
 	}
 
@@ -838,6 +839,7 @@ xilinx_drm_plane_probe_manager(struct drm_device *drm)
 	struct xilinx_drm_plane_manager *manager;
 	struct device *dev = drm->dev;
 	struct device_node *sub_node;
+	const char *format;
 	int ret;
 
 	manager = devm_kzalloc(dev, sizeof(*manager), GFP_KERNEL);
@@ -851,6 +853,20 @@ xilinx_drm_plane_probe_manager(struct drm_device *drm)
 		return ERR_PTR(-EINVAL);
 	}
 
+	/* check the base pixel format of plane manager */
+	ret = of_property_read_string(manager->node, "xlnx,pixel-format",
+				      &format);
+	if (ret < 0) {
+		DRM_ERROR("failed to get a plane manager format\n");
+		return ERR_PTR(ret);
+	}
+
+	ret = xilinx_drm_format_by_name(format, &manager->format);
+	if (ret < 0) {
+		DRM_ERROR("invalid plane manager format\n");
+		return ERR_PTR(ret);
+	}
+
 	manager->drm = drm;
 
 	/* probe an OSD. proceed even if there's no OSD */
-- 
1.7.5.4

